### PR TITLE
feat: add generics to screen queries

### LINKED ISSUE #1033: Make `screen` query methods generic
- `@testing-library/dom` version: 8.5.0
- Testing Framework and version: jest@27.2.0
  
- DOM Environment: jsdom
  

### Relevant code or config:

```ts
import { screen } from '@testing-library/dom';

screen.getByText<HTMLElement>('test');
```

### What you did:

Defined a type argument for `getByText`

### What happened:

```
> tsc

test.ts:3:18 - error TS2558: Expected 0 type arguments, but got 1.

3 screen.getByText<HTMLElement>('test');
                   ~~~~~~~~~~~
```

### Reproduction:

```ts
import { screen } from '@testing-library/dom';

screen.getByText<HTMLElement>('test');
```

### Problem description:

While the methods are generic, the way they're redefined on the `screen` object for example makes them lose their genericness.
Original request: https://github.com/testing-library/dom-testing-library/issues/615
PR which made the methods generic: https://github.com/testing-library/dom-testing-library/pull/1023

### PR DESCRIPTION
**Pull request by timdeschryver**
**What**:

Adding generics to screen queries.
Follow up to #1023, which added generics to query methods
Closes #1033

**Why**:

So users that use screen queries can also use generics.

```ts
screen.getByRole('button') as HTMLButtonElement
// becomes
screen.getByRole<HTMLButtonElement>('button')
```

**How**:

- Added a type test and saw it fail.
- Refactored `BoundFunction`  to a generic type that removes the first argument (the container)
- Manually added all of the queries to the `screen` type because it isn't possible to reuse the queries while keeping the generic

**Checklist**:

- [ ] Documentation added to the
      [docs site](https://github.com/testing-library/testing-library-docs)
- [x] Tests
- [x] TypeScript definitions updated
- [] Ready to be merged

### CODE REVIEW

File: types/get-queries-for-element.d.ts [hunk 1]
```diff
@@ -1,23 +1,162 @@
 import * as queries from './queries'
 
 export type BoundFunction<T> = T extends (
-  attribute: string,
-  element: HTMLElement,
-  text: infer P,
-  options: infer Q,
+  container: HTMLElement,
+  ...args: infer P
```
Comment (timdeschryver) [inline types/get-queries-for-element.d.ts:5]: This syntax is available from TS v3.0.1

**Comment by nstepien** [comment]
@timdeschryver Could you check if the generics work with `within()` as well?

**Comment by timdeschryver** [comment]
@nstepien Yep, we can do that in a follow up PR when this PR is approved.

**Review by eps1lon — APPROVED** [review]

**Comment by eps1lon** [comment]
@timdeschryver Feel free to merge.

**Pull request merged**